nixos/tests/zfs-tests.nix

```diff
@@ -0,0 +1,221 @@
+        pkgs.file
+        pkgs.sysstat # iostat
+        pkgs.pax
+        pkgs.binutils # strings
+        pkgs.parted
+        pkgs.linuxPackages.perf
+        pkgs.samba # net
+        #pkgs.fio # disable because of kernel crashes... / hangs
+        # exportfs -> nfs-utils?
+      ];
+      networking.hostId = "abcdef01";
+    };
+
+  testScript = ''
+      start_all()
+      machine.wait_for_unit("multi-user.target")
+      machine.succeed("modprobe zfs", "zfs version", "id")
+
+      # Running all tests takes very long (multiple hours) and doesn't work properly.
+      # We'll just do a quick smoke-test.
+
+      # test run results (no extra text means PASS):
+      # - chattr
+      # - checksum
+      # - deadman
+      # - devices
+      # - fallocate
+      # - snapshot
+      # - tmpfile
+      # - upgrade
+      # - userquota
+      # - zdb
+      # - zfs
+      # - zfs_sysfs
+      # - zpool_add
+      # - zpool_reopen
+      # - compression (4 FAIL)
+      #   - compression/l2arc_compressed_arc
+      #   - compression/l2arc_compressed_arc_disabled
+      #   - compression/l2arc_encrypted
+      #   - compression/l2arc_encrypted_no_compressed_arc
+      # - events (3 FAIL)
+      #   - events/events_001_pos
+      #   - events/events_002_pos
+      #   - events/zed_rc_filter
+      # - exec (1 FAIL)
+      #   - exec/exec_001_pos
+      # - fault (takes very long, doesn't terminate?)
+      # - large_dnode (1 FAIL)
+      #   - features/large_dnode/large_dnode_006_pos
+      # - misc (1 FAIL)
+      #   - cli_user/misc/arc_summary_001_pos
+      # - mmp (14 SKIP)
+      # - rsend (takes very long, errors out with ENOSPC)
+      # - user_namespace (1 FAIL)
+      # - zfs_mount (2 FAIL)
+      #   - cli_root/zfs_mount/zfs_mount_013_pos
+      #   - cli_root/zfs_mount/zfs_multi_mount
+      # - zfs_share (2 FAIL, 13 SKIP)
+      #   - cli_root/zfs_share/setup
+      # - zpool_create (1 FAIL)
+      #   - cli_root/zpool_create/zpool_create_014_neg
+      # - zpool_expand (1 FAIL)
+      #   - cli_root/zpool_expand/zpool_expand_001_pos
+      # - zpool_split (1 SKIP)
+      #   - cli_root/zpool_split/zpool_split_props
+
+      # TODO
+      # - io (broken? fio)
+      # - pool_checkpoint (takes > 15 minutes)
+      # - cache (fio)
+      # - trim (fio)
+      # - l2arc (fio)
+      # - pam (pamtester)
+
+      tests_to_run = [
+          # 00:07
+          "chattr",
+          # 01:47
+          # "checksum",
+          # 01:00
+          # "deadman",
+          # 00:04
+          "devices",
+          # 00:02
+          "fallocate",
+          # 01:03
+          "snapshot",
+          # 00:02
+          "tmpfile",
+          # 00:22
+          # "upgrade",
+          # 00:58
+          # "userquota",
+          # 04:51
+          # "zdb",
+          # 00:43
+          "zfs",
+          # 00:04
+          "zfs_sysfs",
+          # 04:54
+          # "zpool_add",
+          # 03:06
+          # "zpool_reopen",
+      ]
```

Comment:
If we have to manually specify which tests to run, it'd probably be a good idea to run the longer tests in separate drvs. This way they can be run in parallel/distributed fashion and the total length isn't much of an issue.

I implemented that here: https://github.com/Atemu/nixpkgs/tree/init/zts

Reply:
Looks good, but I am concerned that some of the shorter tests will cause ~20s startup time and then 2 seconds test time. Not sure how bad that is though.

Reply:
We could combine the extremely fast tests into one nixos test but that'd add a good bit of complexity and would seemingly make them all "fail" if only one of them fails. I don't think it's necessary but maybe we could get someone from @NixOS/infra to comment.

Reply:
Yeah, I think just separating is fine.